sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1474,16 +1524,59 @@ _pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
           ZeCommandQueueDesc.ordinal, ZeCommandQueueDesc.index, LowerIndex,
           UpperIndex);
 
-  auto ZeResult = ZE_CALL_NOCHECK(
-      zeCommandQueueCreate, (Queue->Context->ZeContext, Queue->Device->ZeDevice,
-                             &ZeCommandQueueDesc, &ZeQueue));
-  if (ZeResult) {
-    die("[L0] getZeQueue: failed to create queue");
+  // When using immediate commandlists, we create a commandlist corresponding
+  // to the ordinal/index assigned above. Created once, the commandlists are
+  // reused without needing a close or reset. Each commandlist has an implicit
+  // L0 queue.
+  if (UseImmediateCommandLists &&
+      ImmCmdLists[CurrentIndex] == Queue->CommandListMap.end()) {
+    ze_command_list_handle_t ZeCommandList;
+    ZE_CALL_NOCHECK(zeCommandListCreateImmediate,
+                    (Queue->Context->ZeContext, Queue->Device->ZeDevice,
+                     &ZeCommandQueueDesc, &ZeCommandList));
+    ImmCmdLists[CurrentIndex] =
+        Queue->CommandListMap
+            .insert(std::pair<ze_command_list_handle_t, pi_command_list_info_t>{
+                ZeCommandList,
+                {nullptr, false, ZeQueues[CurrentIndex], *QueueGroupOrdinal}})
+            .first;
+    // Add this commandlist to the cache so it can be destroyed as part of
+    // QueueRelease
+    auto &ZeCommandListCache =
+        QueueType != queue_type::Compute
+            ? Queue->Context->ZeComputeCommandListCache[Queue->Device->ZeDevice]
+            : Queue->Context->ZeCopyCommandListCache[Queue->Device->ZeDevice];
+    std::lock_guard<std::mutex> lock(Queue->Context->ZeCommandListCacheMutex);
+    ZeCommandListCache.push_back(ZeCommandList);
+  } else {
+    auto ZeResult =
+        ZE_CALL_NOCHECK(zeCommandQueueCreate,
+                        (Queue->Context->ZeContext, Queue->Device->ZeDevice,
+                         &ZeCommandQueueDesc, &ZeQueue));
+    if (ZeResult) {
+      die("[L0] getZeQueue: failed to create queue");
+    }
   }
+
   return ZeQueue;
 }
 
+// This function will return one of possibly multiple available
+// immediate commandlists associated with this Queue.
```

Comment:
It should be single immediate command-list now

Reply:
For compute operations, there could be any of 4 commandlists created with the compute ordinal and index 0..3. We have gone back and forth on this. The L0 folks prefer using all compute indices although using just 0 will also be OK. For copy operations this could return an ordinal/index of any of the copy engines available.